cycles-ledger/tests/gen.rs

```diff
@@ -0,0 +1,762 @@
+                    .ok_or("total supply underflow")?;
+            }
+            CyclesLedgerCallArg::TransferFrom(TransferFromArgs {
+                spender_subaccount,
+                from,
+                to,
+                amount,
+                fee,
+                memo,
+                created_at_time,
+            }) => {
+                self.execute(&CyclesLedgerCall {
+                    caller: from.owner,
+                    arg: CyclesLedgerCallArg::Transfer(TransferArg {
+                        from_subaccount: from.subaccount,
+                        to: *to,
+                        fee: fee.to_owned(),
+                        created_at_time: *created_at_time,
+                        memo: memo.to_owned(),
+                        amount: amount.clone(),
+                    }),
+                })?;
+                let spender = Account {
+                    owner: arg.caller,
+                    subaccount: *spender_subaccount,
+                };
+                let old_allowance = self
+                    .allowances
+                    .get(&(*from, spender))
+                    .unwrap_or_else(|| panic!("Allowance of {:?} is 0", (from, spender)));
+                let amount = amount.0.to_u128().ok_or("amount is not a u128")?;
+                let new_allowance = old_allowance
+                    .checked_sub(amount)
+                    .and_then(|b| b.checked_sub(FEE))
+                    .ok_or("allowance underflow")?;
+                self.allowances.insert((*from, spender), new_allowance);
+            }
+        }
+        Ok(())
+    }
+}
+
+// Represents a set of valid calls and the
+// in-memory state that results when performing those
+// calls.
+#[derive(Clone, Debug, Default)]
+pub struct CyclesLedgerCallsState {
+    pub calls: Vec<CyclesLedgerCall>,
+    pub state: CyclesLedgerInMemory,
+}
+
+impl CyclesLedgerCallsState {
+    fn new(depositor_cycles: u128) -> Self {
+        Self {
+            calls: vec![],
+            state: CyclesLedgerInMemory::new(depositor_cycles),
+        }
+    }
+
+    // Return the number of tokens available for minting
+    fn token_pool(&self) -> u128 {
+        self.state.token_pool()
+    }
+
+    fn accounts_with_at_least_fee(&self) -> Vec<(Account, u128)> {
+        self.state
+            .balances
+            .iter()
+            .filter_map(|(account, balance)| {
+                if balance >= &FEE {
+                    Some((*account, *balance))
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+}
+
+impl IsCyclesLedger for CyclesLedgerCallsState {
+    fn execute(&mut self, arg: &CyclesLedgerCall) -> Result<(), String> {
+        self.state.execute(arg)?;
+        self.calls.push(arg.clone());
+        Ok(())
+    }
+}
+
+fn arb_allowed_principal() -> impl Strategy<Value = Principal> {
+    collection::vec(any::<u8>(), 0..30).prop_filter_map(
```

Comment:
technically the last byte is not allowed to be an arbitrary number, but I guess until Principal starts to check that it's a case of 'good enough'

Reply:
If we want it to be better than "good enough", is there an easy fix?

Reply:
I'm also interested in knowing more. Can we make the generator more precise?